### ISSUE TITLE
Audit logs: student list is empty

### ISSUE THREAD

**Issue by damithc** [TEAMMATES/teammates#11122]
Steps:
1. Instructor goes to the audit logs page using `/instructor/logs` ~~Admin goes to the audit logs page of an instructor using the masquerade mode~~
2. Selects a course ID
3. Problem: student list dropdown is not populated

![image](https://user-images.githubusercontent.com/1673303/116181580-6ee8c100-a74d-11eb-8e0f-cf6dc722aff8.png)

**Comment by damithc** [comment]
In fact, this doesn't work even in the normal mode.

**Comment by damithc** [comment]
@t-cheepeng @AdithyaNarayan @halfwhole any ideas?

**Comment by wkurniawan07** [comment]
Not reproducible.

**Comment by damithc** [comment]
Seems to work now. Not sure why it didn't work. It didn't work on the day I posted this issue and it didn't work when I tried it again a few days ago. 🤔

**Comment by t-cheepeng** [comment]
Could not reproduce this. Was the previous course archived or deleted?

**Comment by damithc** [comment]
Happening again:
![image](https://user-images.githubusercontent.com/1673303/117604810-e6c0dd80-b188-11eb-9f8c-0d0980e33cd2.png)

Refreshed the page and it started working. Some sort of caching issue? Or a race condition in populating the dropdown?

**Comment by t-cheepeng** [comment]
Maybe could be a caching issue? Try to clear browser caches and see if the problem persists. There shouldn't be any race conditions in the retrieval of list of students

**Comment by damithc** [comment]
> Maybe could be a caching issue? Try to clear browser caches and see if the problem persists. There shouldn't be any race conditions in the retrieval of list of students

It still happens intermittently. Note that I have a lot of courses (and several thousand students). How are the student names retrieved? All at once?

**Comment by t-cheepeng** [comment]
Yes, the student names are all retrieved at once for all active courses

**Comment by rrtheonlyone** [comment]
I am not sure why we choose to retrieve all student names for all active courses on page load. I feel it makes more sense to retrieve the students (and cache it) once the dropdown course is selected. @damithc shall we try this?

After that, we can see if the above error still happens. It is difficult to re-create the above error because it might be a result of the number of students prof has. (I have only 20 students on my setup at the moment)

**Comment by damithc** [comment]
Still happens around 30% of the time. Even when using a new computer. Here is an example (not a definite way to reproduce though):

Load the page. Seems to work.
Go to another page.
Go back to the previous page again. Doesn't work now.

> I am not sure why we choose to retrieve all student names for all active courses on page load. I feel it makes more sense to retrieve the students (and cache it) once the dropdown course is selected. @damithc shall we try this?

Given this feature is not expected to be used often, I don't mind a simpler solution. We can go for on-demand retrieval if the additional development cost is not too high.


> After that, we can see if the above error still happens. It is difficult to re-create the above error because it might be a result of the number of students prof has. (I have only 20 students on my setup at the moment)

Yes, might be related to the total students count.

**Comment by rrtheonlyone** [comment]
> Given this feature is not expected to be used often, I don't mind a simpler solution. We can go for on-demand retrieval if the additional development cost is not too high.

It is not difficult to do `on-demand` retrieval. I am guessing that should resolve the above problem too. Let me try my hand at tackling this.

**Comment by damithc** [comment]
I was trying this feature today (in masquerade mode) and encountered the problem again. That instructor had a course with 1300 student, in case that matters.

**Comment by damithc** [comment]
To add: Another user reported that he faces this problem for larger courses but it works for smaller courses.

**Comment by madanalogy** [comment]
Reopening as resolving PR has been reverted

**Comment by halfwhole** [comment]
Strangely, it looks to me like this issue has already been resolved - the original PR (#11479) doesn't look like it's been reverted (can't find the revert commit in history), and the PR changes are still currently in master too. Could I just sanity check with @madanalogy that this issue can be closed!

**Comment by madanalogy** [comment]
Yeap you're right, I reverted the revert but forgot to close the issue again. Thanks for the catch @halfwhole
